azure-pipelines-pr.yml

```diff
@@ -0,0 +1,331 @@
+      sourceRepositoriesToScan:
+        exclude:
+          - repository: forwardCompatibleRelease
+    pool:
+      name: Cosmos2MT-AzureDevOps-AgentPool
+      image: 1es-pt-windows-2019
+      os: windows
+    customBuildTags:
+      - ES365AIMigrationTooling
+          
+    stages:
+    - stage: Build
+      displayName: Build Sources
+      jobs:
+      - job: Build
+        templateContext:
+          outputs:
+            - output: pipelineArtifact
+              artifactName: Microsoft.Spark.Binaries
+              targetPath: $(Build.ArtifactStagingDirectory)
+              displayName: '[PUBLISH] Microsoft.Spark.Binaries'
+
+        steps:
+        - task: Maven@3
+          displayName: 'Maven build src'
+          inputs:
+            mavenPomFile: src/scala/pom.xml
+
+        - task: Maven@3
+          displayName: 'Maven build benchmark'
+          inputs:
+            mavenPomFile: benchmark/scala/pom.xml
+
+        - script: build.cmd -pack
+                    -c $(buildConfiguration)
+                    -ci
+                    $(_OfficialBuildIdArgs)
+                    /p:PublishSparkWorker=true
+                    /p:SparkWorkerPublishDir=$(Build.ArtifactStagingDirectory)\Microsoft.Spark.Worker
+          displayName: '.NET build'
+
+        - task: DotNetCoreCLI@2
+          displayName: '.NET unit tests'
+          inputs:
+            command: test
+            projects: '**/*UnitTest/*.csproj'
+            arguments: '--configuration $(buildConfiguration)'
+
+        - task: CopyFiles@2
+          displayName: Stage Maven build jars
+          inputs:
+            sourceFolder: $(Build.SourcesDirectory)/src/scala
+            contents: '**/*.jar'
+            targetFolder: $(Build.ArtifactStagingDirectory)/Jars
+
+    - template: azure-pipelines-e2e-tests-template.yml
+      parameters:
+        backwardCompatibleRelease: $(backwardCompatibleRelease)
+        forwardCompatibleRelease: $(forwardCompatibleRelease)
+        tests:
+        - version: '2.4.0'
+          enableForwardCompatibleTests: false
+          enableBackwardCompatibleTests: false
+          jobOptions:
+          - pool: 'Windows'
+            testOptions: ""
+            backwardCompatibleTestOptions: $(backwardCompatibleTestOptions_Windows_2_4)
+            forwardCompatibleTestOptions: $(forwardCompatibleTestOptions_Windows_2_4)
+          - pool: 'Linux'
+            testOptions: ""
+            backwardCompatibleTestOptions: $(backwardCompatibleTestOptions_Linux_2_4)
+            forwardCompatibleTestOptions: $(forwardCompatibleTestOptions_Linux_2_4)
+        - version: '2.4.1'
+          enableForwardCompatibleTests: false
+          enableBackwardCompatibleTests: false
+          jobOptions:
+          - pool: 'Windows'
+            testOptions: ""
+            backwardCompatibleTestOptions: $(backwardCompatibleTestOptions_Windows_2_4)
+            forwardCompatibleTestOptions: $(forwardCompatibleTestOptions_Windows_2_4)
```

Comment:
We can use foreach here, to reduce duplication: 
https://learn.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#each-keyword

Reply:
updated.